### ISSUE TITLE
Include static list of kernel firmware files to add to Debian rootfs images

### ISSUE THREAD

**Issue by gctucker** [kernelci/kernelci-core#698]
A first step to start using kernel firmware files is to define a static list with the ones to copy from the upstream repository into rootfs images, such as buster-v4l2.  This one is used as a ramdisk, so the size should be kept to only the firmware files that may be used on platforms currently running v4l2-compliance tests in KernelCI.

Typically, the [buster-v4l2.sh script](https://github.com/kernelci/kernelci-core/blob/main/config/rootfs/debos/scripts/buster-v4l2.sh) would add a step to check out the [linux-firmware](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git) repository and copy files.

This could then be generalised with a separate script that can be used in any rootfs image, with a list of firmware files to copy defined in [rootfs-configs.yaml](https://github.com/kernelci/kernelci-core/blob/main/config/core/rootfs-configs.yaml) in the same way that Debian packages and files to remove are listed there.

**Comment by eballetbo** [comment]
Some examples of missing firmware:

[baseline-mt8173-elm-hana-next-2021-05-27](https://storage.kernelci.org/next/master/next-20210527/arm64/defconfig/clang-12/lab-collabora/baseline-mt8173-elm-hana.html)
- [   14.823968] r8152 2-1.1:1.0: Direct firmware load for rtl_nic/rtl8153b-2.fw failed with error -2
- [   14.833068] r8152 2-1.1:1.0: Falling back to sysfs fallback for: rtl_nic/rtl8153b-2.fw

[asus-C433TA-AJ0005-rammus](https://lava.collabora.co.uk/scheduler/job/3483318)

- [    0.892283] i915 0000:00:02.0: Direct firmware load for i915/kbl_dmc_ver1_04.bin failed with error -2
- [    0.893101] i915 0000:00:02.0: [drm] Failed to load DMC firmware i915/kbl_dmc_ver1_04.bin. Disabling runtime power management.
- [    5.049971] r8152 2-3:1.0: Direct firmware load for rtl_nic/rtl8153b-2.fw failed with error -2
- [    5.059940] r8152 2-3:1.0: unable to load firmware patch rtl_nic/rtl8153b-2.fw (-2)
- [   10.906283] platform regulatory.0: Direct firmware load for regulatory.db failed with error -2

**Comment by gctucker** [comment]
I'm not sure we can easily add firmware files to the baseline rootfs, as it needs to be tiny and it's used on all the devices.  So maybe focusing just on buster-v4l2 or other more specialised rootfs variants would be easier to start with.

**Comment by eballetbo** [comment]
This job downloads a rootfs containing some firmware needed by octopus: https://lava.collabora.co.uk/scheduler/job/3943690

You can see in dmesg that the firmware is loaded
```
<6>[    3.611852] i915 0000:00:02.0: [drm] Finished loading DMC firmware i915/glk_dmc_ver1_04.bin (v1.4)
```

This is simply hacking the rootfs script generation with the following:
```
########################################################################
# Install firmware files                                               #
########################################################################

# Missing firmware for octopus
FW_LIST="\
    rtl_nic/rtl8153b-2.fw \
    i915/kbl_dmc_ver1_04.bin \
    i915/glk_dmc_ver1_04.bin \
"

FW_URL="git://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git"

mkdir -p /tmp/linux-firmware && cd /tmp/linux-firmware
git clone --depth=1 $FW_URL .

mkdir -p /lib/firmware

for file in ${FW_LIST}; do cp -v --parents ${file} /lib/firmware/ ; done
```

**Comment by gctucker** [comment]
I've re-run the job with `ls -l /lib/firmware`: https://lava.collabora.co.uk/scheduler/job/3952700#L2513

This looks good, thanks.  It could be added to the debos `rootfs.yaml` base template with a script to copy a list of firmware files with an extra template variable in the same way that we have `extra_files_remove` etc...  This would then be configured in `rootfs-configs.yaml` so we could have variants of the existing rootfs images with specific firmware files added to them.  We may also have an NFS image with all the firmware files, which would be very large but could be useful for prototyping new tests.

A longer-term solution would probably be to have the ability to provide a tarball with some firmware files determined just before scheduling each job, in the same way that we have a tarball for kernel modules.  That way we could include files depending on the target device type, the kernel build, tests to run... and avoid making special rootfs images.  It's a bit more complicated to implement though.

**Comment by gctucker** [comment]
Moving to hackfest 2 board.

**Comment by 10ne1** [comment]
FYI: @gctucker & @eballetbo I implemented in PR #801 a new command which builds device-specific firmware tarballs which then could be uploaded & used on various devices as needed. This last uploading & using part still needs to be implemented if you're ok with my approach. In a nusthell I wanted to directly go with the "long term" design described by @gctucker above. Thanks!

**Comment by 10ne1** [comment]
To summarize the current status:

 A comparison between the two proposed firmware integration designs revealed that the "long-term" approach suggested above with separate tarballs is less desirable than installing a predetermined fw subset directly in the rootfs which is simpler and more robust, so we should go with the first approach.
